Sources/AsyncHTTPClient/RedirectState.swift

```diff
@@ -0,0 +1,134 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2018-2019 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import struct Foundation.URL
+import NIOHTTP1
+
+struct RedirectState {
+    /// number of redirects we are allowed to follow
+    private var count: Int
+
+    /// nil if we should ignore cycles, otherwise a set of all visited URLs
+    private var visited: Set<URL>?
+}
+
+extension RedirectState {
+    /// Creates a `RedirectState` from a configuration.
+    /// Returns nil if the user disallowed redirects,
+    /// otherwise an instance of `RedirectState` which respects the user defined settings.
+    init?(_ configuration: HTTPClient.Configuration.RedirectConfiguration.Configuration) {
+        switch configuration {
+        case .disallow:
+            return nil
+        case .follow(let maxRedirects, let allowCycles):
+            self.init(count: maxRedirects, visited: allowCycles ? nil : [])
+        }
+    }
+}
+
+extension RedirectState {
+    /// Call this method when you are about to do a redirect to the given `redirectURL`.
+    /// This method records that URL into `self`.
```

Comment:
This only happens sometimes. As mentioned above I think we should always record.